lib/src/rules/prefer_is_not_operator.dart

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'Prefer is! operator.';
+
+const _details = r'''
+When checking if the object has not specified type, it is preferable to use 'is!' operator.
+
+**BAD:**
+```
+if (!(foo is Foo)) {
+  ...
+}
+```
+
+**GOOD:**
+```
+if (foo is! Foo) {
+  ...
+}
+```
+
+''';
+
+class PreferIsNotOperator extends LintRule implements NodeLintRule {
+  PreferIsNotOperator()
+      : super(
+            name: 'prefer_is_not_operator',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this);
+    registry.addIsExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitIsExpression(IsExpression node) {
+    // Check whether is expression is inside parenthesis
+    if (node.parent is ParenthesizedExpression) {
+      final isExpressionInsideParenthesis = node.parent;
+      // Check for NOT (!) operator
+      if (isExpressionInsideParenthesis.parent.beginToken.type ==
+          TokenType.BANG) {
+        rule.reportLint(isExpressionInsideParenthesis.parent);
+      }
+    }
```

Comment:
I think this may be more robust?

```dart
    final prefixExpression =
        node.parent.thisOrAncestorOfType<PrefixExpression>();
    if (prefixExpression?.operator?.type == TokenType.BANG) {
      rule.reportLint(prefixExpression);
    }
```

(Deferring to @scheglov and @bwilkerson who know the AST model better...)

Reply:
I'd go with pure `.parent` checks.
Something like (not tested):
```dart
if (node.notOperator == null) {
  final parent = node.parent;
  if (parent is ParenthesizedExpression) {
    final parent2 = parent.parent;
    if (parent2 is PrefixExpression && parent2.operator.type = TokenType.BANG) {
      ...report
    }
  }
}
```

Reply:
Yes checking for `is!` in the beginning is better. 

And I also prefer checking for `ParanthesizedExpression` before doing NOT operator check for the parenthesized expression. 

```dart
  void visitIsExpression(IsExpression node) {
    // Return if it is `is!` expression
    if(node.notOperator != null) {
      return;
    }

    // Check whether is expression is inside parenthesis
    if (node.parent is ParenthesizedExpression) {
      final isExpressionInsideParenthesis = node.parent;
      final prefixExpression = isExpressionInsideParenthesis.parent;
      // Check for NOT (!) operator
      if (prefixExpression is PrefixExpression &&
          prefixExpression?.operator?.type == TokenType.BANG) {
        rule.reportLint(isExpressionInsideParenthesis.parent);
      }
    }
  }
```

Let me know if you have any comments, so that I can update the PR.

Reply:
SGTM!

Reply:
The name `isExpressionInsideParenthesis` does not sound right. It is not "inside" parenthesis, it is parenthesis. Maybe just `parenthesizedExpression`.

We have already checked that `prefixExpression is PrefixExpression`, so it cannot be `null`, so `prefixExpression?.operator` is not necessary, just `prefixExpression.operator` will work. Correspondingly the second `?.type` also does not have to be null aware.